File: frontend/src/stores/analysisScopes.ts
import { defineStore } from 'pinia'

import { apiGet, apiRequest } from '@/api/client'
import type { AnalysisScopesViewDocument } from '@/types/api'

export interface AnalysisScopeInput {
  name: string
  purpose: string
  paper_ids: string[]
  source_filter_snapshot: Record<string, string>
}

export const useAnalysisScopeStore = defineStore('analysis-scopes', {
  actions: {
    list(projectId: string) {
      return apiGet<AnalysisScopesViewDocument>(`/projects/${projectId}/analysis-scopes`)
    },
    create(projectId: string, input: AnalysisScopeInput, expectedRevision: number) {
      return apiRequest<AnalysisScopesViewDocument>(`/projects/${projectId}/analysis-scopes`, {
        method: 'POST',
        body: { ...input, expected_revision: expectedRevision },
      })
    },
    update(
      projectId: string,
      scopeId: string,
      input: AnalysisScopeInput,
      expectedRevision: number,
    ) {
      return apiRequest<AnalysisScopesViewDocument>(
        `/projects/${projectId}/analysis-scopes/${scopeId}`,
        {
          method: 'PATCH',
          body: { ...input, expected_revision: expectedRevision },
        },
      )
    },
    remove(projectId: string, scopeId: string, expectedRevision: number) {
      return apiRequest<AnalysisScopesViewDocument>(
        `/projects/${projectId}/analysis-scopes/${scopeId}`,
        {
          method: 'DELETE',
          body: { expected_revision: expectedRevision },
        },
      )
    },
  },
})

File: backend/src/papermatrix/api/v1/analysis_scopes.py
"""Reproducible analysis-scope API."""

from __future__ import annotations

from typing import cast
from uuid import UUID

from fastapi import APIRouter, Request, status
from pydantic import BaseModel, Field

from papermatrix.domain.analysis_scope import AnalysisScopesViewDocument
from papermatrix.services.analysis_scope_service import AnalysisScopeService

router = APIRouter(tags=["analysis-scopes"])


def _service(request: Request) -> AnalysisScopeService:
    return cast(AnalysisScopeService, request.app.state.analysis_scope_service)


class AnalysisScopeInput(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    purpose: str = Field(default="", max_length=5000)
    paper_ids: list[UUID] = Field(min_length=1)
    source_filter_snapshot: dict[str, str] = Field(default_factory=dict)
    expected_revision: int = Field(ge=0)


class AnalysisScopeUpdate(AnalysisScopeInput):
    expected_revision: int = Field(ge=1)


class AnalysisScopeDelete(BaseModel):
    expected_revision: int = Field(ge=1)


@router.get(
    "/projects/{project_id}/analysis-scopes",
    response_model=AnalysisScopesViewDocument,
)
def list_analysis_scopes(
    project_id: UUID,
    request: Request,
) -> AnalysisScopesViewDocument:
    return _service(request).list_scopes(project_id)


@router.post(
    "/projects/{project_id}/analysis-scopes",
    response_model=AnalysisScopesViewDocument,
    status_code=status.HTTP_201_CREATED,
)
def create_analysis_scope(
    project_id: UUID,
    payload: AnalysisScopeInput,
    request: Request,
) -> AnalysisScopesViewDocument:
    return _service(request).create(project_id, **payload.model_dump())


@router.patch(
    "/projects/{project_id}/analysis-scopes/{scope_id}",
    response_model=AnalysisScopesViewDocument,
)
def update_analysis_scope(
    project_id: UUID,
    scope_id: UUID,
    payload: AnalysisScopeUpdate,
    request: Request,
) -> AnalysisScopesViewDocument:
    return _service(request).update(project_id, scope_id, **payload.model_dump())


@router.delete(
    "/projects/{project_id}/analysis-scopes/{scope_id}",
    response_model=AnalysisScopesViewDocument,
)
def delete_analysis_scope(
    project_id: UUID,
    scope_id: UUID,
    payload: AnalysisScopeDelete,
    request: Request,
) -> AnalysisScopesViewDocument:
    return _service(request).delete(project_id, scope_id, **payload.model_dump())
